core/embed/trezorhal/stm32f4/usb/usbd.c

```diff
@@ -189,50 +212,67 @@ secbool usb_configured(void) {
   return secfalse;
 }
 
-/*
- * Utility functions for USB interfaces
- */
+// ==========================================================================
+// Utility functions for USB class drivers
+// ==========================================================================
+
+void *usb_get_iface_state(uint8_t iface_num, const USBD_ClassTypeDef *class) {
+  usb_driver_t *usb = &g_usb_driver;
 
-static usb_iface_t *usb_get_iface(uint8_t iface_num) {
   if (iface_num < USBD_MAX_NUM_INTERFACES) {
-    return &usb_ifaces[iface_num];
-  } else {
-    return NULL;  // Invalid interface number
+    usb_iface_t *iface = &usb->ifaces[iface_num];
+
+    if (iface->class == class) {
+      return &iface->state;
+    }
   }
+
+  // Invalid interface number or type
+  return NULL;
 }
 
-static void *usb_desc_alloc_iface(size_t desc_len) {
-  if (usb_config_desc->wTotalLength + desc_len < USB_MAX_CONFIG_DESC_SIZE) {
-    return usb_next_iface_desc;
-  } else {
-    return NULL;  // Not enough space in the descriptor
+void usb_set_iface_class(uint8_t iface_num, const USBD_ClassTypeDef *class) {
+  usb_driver_t *usb = &g_usb_driver;
+
+  if (iface_num < USBD_MAX_NUM_INTERFACES) {
+    if (usb->ifaces[iface_num].class == NULL && class != NULL) {
+      usb->config_desc->bNumInterfaces++;
+    } else if (usb->ifaces[iface_num].class != NULL && class == NULL) {
+      usb->config_desc->bNumInterfaces--;
```

Comment:
not sure how this unsetting works, wouldn't this break e.g. wTotalLength and expected position of the interface in the desc buffer?

Reply:
Yes, this setting was not necessary to add at all as nobody is using it. I added it just for completeness, even though no mechanism for removing classes (deallocating descriptors) has been implemented yet. I was a little bit concerned that it might lead to confusion, but I don't think it's harmful.

Reply:
I understand that it isn't called, so no real problem arises, but if someone calls this the descriptors will be in inconsistent state: regardless which index of interface you unset, the host will see the first few added. So unless the descriptor deallocation is properly implemented in the latter PR, which i havent seen yet, i would be for removing this functionality

Reply:
Ok, no problem with this. We may get it back in the future if needed. See d013de692c5f7358ea4d96a9c393160671daa6c8